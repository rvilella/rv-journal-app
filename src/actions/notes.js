import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';

// react-journal


export const startNewNote = () => {
    return async(dispatch, getState) => {

        const uid = getState().auth.uid;
       
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
        
            dispatch(activeNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));

        } catch (error) {
            console.log(error);
        }

    }    
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {

        const {uid} = getState().auth;

        if(!note.url){
            delete note.url;
        }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        dispatch(refresNote(note.id, note));

        Swal.fire('Saved', note.title, 'success');

    }
}

export const refresNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState) => {
        // const activeNote = getState().notes.active;
        const {active: activeNote} = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            onBeforeOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        // console.log(fileUrl);
        dispatch(startSaveNote(activeNote))

        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async(dispatch, getState) => {
        
        const uid = getState().auth.uid;
        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));

    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
})