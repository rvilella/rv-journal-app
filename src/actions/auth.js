import Swal from 'sweetalert2';

import { types } from '../types/types';
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) =>{

        // setTimeout(() => {
            
        //     dispatch(login(123, 'Pedro'));

        // }, 3500);

        dispatch(startLoading());

        return firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        })
        .catch(e => {
            console.log(e);
            // acÃ¡ se ejecuta por si el usuario escribe algo mal
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        })

    }
}

export const startRegisterWithNameEmailPassword = (name, email, password) => {

    return (dispatch) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async({user}) => {

                await user.updateProfile({displayName: name})

                dispatch(
                    login(user.uid, user.displayName)
                );
            })
            .catch(e => {
                console.log(e);
                Swal.fire('Error', e.message, 'error')
            })
    }

}

export const startGoogleLogin = () => {
    return (dispatch) => {

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
               )
            });

    }
}

export const login = (uid, displayName) => {
    return{
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const startLogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());

        dispatch(noteLogout());
    }
}

export const logout = () => ({
    type: types.logout
})
