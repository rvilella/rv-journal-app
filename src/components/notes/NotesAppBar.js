import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();

    const {active} = useSelector(state => state.notes);

    const handleSave = () => {
        dispatch(startSaveNote(active));
    }

    const handlePictureClick = () => {
        // # --> busca por el id
        document.querySelector('#fileSelector').click();
    }

    const handleFileChange = (e) => {
        // console.log(e);
        const file = e.target.files[0];
        if (file) {
            dispatch(startUploading(file));
            // para que se resetee el valor y permita subir una misma imagen a 2 notas
            document.querySelector('#fileSelector').value = '';
        }
    }

    return (
        <div className="notes__appbar">
            <span>14 de enero de 2021</span>

            <input
                id="fileSelector"
                type="file"
                name="file"
                style={{display: 'none'}}
                onChange={handleFileChange}
            />

            <div>
                <button 
                    className="btn"
                    onClick={handlePictureClick}
                >
                    Picture
                </button>

                <button 
                    className="btn"
                    onClick={handleSave}
                >
                    Save
                </button>
            </div>
        </div>
    )
}
