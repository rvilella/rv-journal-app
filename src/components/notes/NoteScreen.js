import React, { useEffect, useRef } from 'react';
import Swal from 'sweetalert2';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const {active: note} = useSelector(state => state.notes);
    const [formValues, handleInputChange, reset] = useForm(note);
    const {title, body, id} = formValues;

    // almacena una variable mutable que no va a renderizar todo el compente si cambia
    const activeId = useRef(note.id);

    // referencia al url de la imagen, para que cuando cambie vuelva a resetear el estado de la nota
    const activeUrl = useRef(note.url)

    useEffect(() => {
        
        // evalua si la nota activa es distinta de la almacenada previamente
        // con current se evalua el valor actual
        if(note.id !== activeId.current) {
            // se redibuja el formulario con la nueva nota activa
            reset(note);
            // se establece el valor actual del activeId
            activeId.current = note.id;
        }
        if(note.url !== activeUrl.current) {
            reset(note);
            activeUrl.current = note.url;
        }

    }, [note, reset])

    useEffect(() => {
       
        dispatch(activeNote(formValues.id, {...formValues}));

    }, [formValues, dispatch])

    const handleDelete = () => {
        dispatch(startDeleting(id));
        Swal.fire('Deleted', note.title, 'success')
    }

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">

                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value={title} 
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder="What happened today"
                    className="notes__textarea"
                    autoComplete="off"
                    name="body"
                    value={body}
                    onChange={handleInputChange}
                ></textarea>

                {        
                
                    (note.url) &&
                    <div className="notes__image">
                        <img
                            src={note.url}
                            alt="imagen"
                        />

                    </div>
                }

            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>

        </div>
    )
}
