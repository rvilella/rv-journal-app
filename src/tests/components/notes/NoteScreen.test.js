import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';


jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));



const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {
        uid: '1',
        name: 'Rodrigo'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: 1234,
            title: 'Hola',
            body: 'Mundo',
            date: 0
        },
        notes: []
    }
};

// se crea un store (redux/state/auth/uid)
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store= {store}>
        <NoteScreen />
    </Provider>
);

describe('Pruebas en <NoteScreen />', () => {
   
    test('debe de mostarse correctamente', () => {
       
        expect(wrapper).toMatchSnapshot();

    });


    test('debe de disparar el activeNote', () => {
        
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Hola de nuevo'
            }
        });

        // expect(activeNote).toHaveBeenCalled();

        // se evalua solamente la ultima llamada
        expect(activeNote).toHaveBeenLastCalledWith(
            1234,
            {
                body: 'Mundo',
                title: 'Hola de nuevo',
                id: 1234,
                date: 0
            }
        );

    });
    
});
