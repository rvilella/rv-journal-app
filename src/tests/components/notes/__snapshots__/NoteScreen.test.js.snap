// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> debe de mostarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      14 de enero de 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
